geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs

```diff
@@ -65,6 +130,277 @@ data MMToken = MMToken
+          , pccPriceDiffThreshold1    = 1
+          , pccPriceDiffThreshold2    = 1
+          , pccRespiteDelayFactor     = 0.5     -- TODO: sensible value?
+          , pccAfterExitRelaxAim1     = 0       -- placeholder, unused
+          , pccAfterExitWorseMax1     = 0       -- placeholder, unused
+          , pccAfterExitRelaxAim2     = 10      -- TODO: sensible value?
+          , pccPricesProvidersWeights = [1]
+          }
+
+        , ppaPricesCluster = [ MaestroConfig
+            { mcApiKey             = mcApiKey
+            , mcResolutionOverride = Just pcResolution
+            , mcDex                = mcDex
+            , mcPairOverride       = mcPairOverride
+            , mcOffsetFilePath     = Nothing  -- TESTING
+            , mcAvailablePath      = Nothing  -- TESTING
+            } ]
+        }
+
+    PCVersion2 (PriceConfigV2 {..}) → do
+      let ppaCommonCfg     = pcPriceCommonCfg
+          thresHold1       = pccPriceDiffThreshold1 ppaCommonCfg
+          thresHold2       = pccPriceDiffThreshold1 ppaCommonCfg
+          ppaPricesCluster = NE.toList pcPricesProviderCfgs
+
+      when (length (pccPricesProvidersWeights ppaCommonCfg) /= length ppaPricesCluster) $
+        throwIO $ userError "Expected 'pccPricesProvidersWeights' to be of same length as 'pcPricesProviderCfgs'."
+      when (thresHold2 < thresHold1) $
+        throwIO $ userError "Expected 'pccPriceDiffThreshold2 >= pccPriceDiffThreshold1'."
+
+      return PPA
+        { ppaCommonCfg     = ppaCommonCfg,
+          ppaPricesCluster = ppaPricesCluster
+        }
+  
+data SourceError = SourceUnavailable
+  deriving stock Show
+
+data PriceIndicator = PriceMismatch1 | PriceMismatch2 | PriceUnavailable | PriceSourceFail Price | PriceAverage Price
+  deriving stock Show
+
+newtype GetQuota = GetQuota { getQuota ∷ MMTokenPair → IO (Either SourceError Price) }
+
+buildGetQuota ∷ PriceCommonCfg → PricesProviderCfg → GetQuota
+
+buildGetQuota PriceCommonCfg {..} MaestroConfig {..} = GetQuota $ \mmtp → do
+  env ← networkIdToMaestroEnv (coerce mcApiKey) pccNetworkId
+  res ← maybe (fromCommonResolution pccCommonResolution) pure mcResolutionOverride
+
+  (pairName, commodityIsA) ← case mcPairOverride of
+    -- We have to override with given details.
+    Just (MaestroPairOverride {..}) → do
+      pure (pack mpoPair, mpoCommodityIsFirst)
+    -- We are given commodity token and need to find pair name.
+    Nothing → do
+      allDexPairs ← dexPairResponsePairs <$> (handleMaestroError (functionLocationIdent <> " - fetching dex pairs") <=< try $ pairsFromDex env mcDex)
+
+      let go []           = throwIO MaestroPairNotFound
+          go (dpi : dpis) = maybe (go dpis) pure $ isRelevantPairInfo mmtp dpi
+      first dexPairInfoPair <$> go allDexPairs
+
+  let pair = TaggedText pairName
+
+  ohlInfo ← handleMaestroError (functionLocationIdent <> " - fetching price from pair") <=< try $ pricesFromDex env mcDex pair (Just res) (Just Descending)
```

Comment:
`handleMaestroError` would be throwing error and I suppose you wanted to return `SourceUnavailable`?

Reply:
Indeed.  I have modified accordingly.